Version2 landcover

tobstac opened the pull request:
Merge request for landcover diagnostic. Also closes #756 . The implementation of the landcover diagnostic performs similar to the landcover diagnostic in the ESMValTool Version 1. However, only the MPI-M contribution (bar plots displaying accumulated area, mean fraction and bias against observations) was implemented. Further diagnostics might be implemented later by @BenMGeo .
There is one unresolved issue  concerning the dimensions of the baresoilFrac Variable. According to definition it should have the dimensions [longitude latitude time typebare] (see esmvaltool/cmor/tables/cmip5/Tables/CMIP5_Lmon). However, all the experiments I downloaded from ESGF have only the dimensions [longitude latitude time] for baresoilFrac. I could work around this by simply changing the definition in the CMIP5_Lmon file, but I guess this is the official definition and should not be changed. What would be the correct way to deal with this?

mattiarighi commented:
Thanks @tobstac :clap: 

Which v1 namelist is this replacing (see list here #503)?

> There is one unresolved issue concerning the dimensions of the baresoilFrac Variable. According to definition it should have the dimensions [longitude latitude time typebare] (see esmvaltool/cmor/tables/cmip5/Tables/CMIP5_Lmon). However, all the experiments I downloaded from ESGF have only the dimensions [longitude latitude time] for baresoilFrac. 

Does the tool read the variable anyway or does it return an error?

> I could work around this by simply changing the definition in the CMIP5_Lmon file, but I guess this is the official definition and should not be changed. 

Right, CMOR tables should not be changed (only the custom ones, but this is not the case here).

> What would be the correct way to deal with this?

You could try implementing a fix, but since it affects **all** datasets it's gonna be nasty.

Another workaround would be to defined a derived variable with a similar name: the derivation function would just do nothing, but it could use a custom CMOR table with the correct dimension names.

mattiarighi commented:
> It replaces the namelist_landcover.xml. It is not included in the list, maybe because it was in the private repository.

Ok, thanks. I need to update #503 with the namelists of the private repo.

>  Is there any guide or example about how to do a fix?

You can follow one of the many [fix files](https://github.com/ESMValGroup/ESMValTool/tree/version2_development/esmvaltool/cmor/_fixes/CMIP5) already available in the code.
You need to create 1 fix file for each dataset (if not already available) and add a class named like the variable you want to fix, which should contain the code to fix the data. 

Note that due to python naming conventions, the dash `-` in dataset names must be replaced by an underscore `_` in the fix file. For example the fix for `CNRM-CM5` is named [CNRM_CM5.py](https://github.com/ESMValGroup/ESMValTool/blob/version2_development/esmvaltool/cmor/_fixes/CMIP5/CNRM_CM5.py).

tobstac commented:
Thanks for the links to the fix files!
I played around with writing a fix file and noticed that - while the variable itself has just the three dimensions - an additional scalar dimension named area_type with the value bare_ground does already exist in the experiment files. That means the information is there, it is just not part of the variable dimensions. That leads to the question whether or not baresoilFrac is really supposed to be a 4D variable (especially considering that area_type has a length of just 1). And following from this, I wonder whether the right way to proceed would really be to write a fix for every model. Could it be that variables like this are kind of a special case which would be needed to taken care of in the general read routines of the esmvaltool?

mattiarighi commented:
@axel-lauer @BenMGeo can you give us feedback?
How was this implemented in v1? I think we used derived variables there.

axel-lauer commented:
In version 1.1, namelist_lauer17rse, we used aggregated variables, i.e. *shrubNtreeFrac* and *grassNcropFrac*. In order to derive these variables, we read *bareSoilFrac*, *grassFrac*, *shrubFrac* and *treeFrac*. Reading in those 4 source variables was not a problem because we did not have such strict variable checks then.

Btw, this namelist was initially not in the list of namelists to be ported to v2 since we never got a pull request for the above mentioned feature branch.

tobstac commented:
I tried to track the error somewhat further and ended up at
https://github.com/ESMValGroup/ESMValTool/blob/5b0ba6ecf59eefa8a092878cc96d866f512789a2/esmvaltool/cmor/check.py#L271-L272
Here, the coordinate 'type' is not found in self._cube.coord as it only contains the dimensions of the cube (time lat and lon) and not the auxillary variable (which is part of self._cmor_var.coordinates.items() used for defining the loop), causing the error. Interestingly, this issue seemed to be anticipated a few lines above:
https://github.com/ESMValGroup/ESMValTool/blob/5b0ba6ecf59eefa8a092878cc96d866f512789a2/esmvaltool/cmor/check.py#L245-L246
where such dimensions are excluded and the error message is just passed. Would it be a possibility to sort dimensions according to whether they are part of the cube dimensions or just auxillary, or would this open the possibility of letting other errors slip through?

tobstac commented:
Progress. I just realized that check.py is not to blame. The typebare dimension is included in MPI-ESM, but not in inmcm4 and, of course, not in the esacci landcover observations. Adding fixes for both, I can run the diagnostics without any need to change the CMIP5_Lmon.

tobstac commented:
Is there anything more I need to do for this? The new issues in codacy are related to the CapWords convention and class name pattern. I guess both cannot be changed as the class name has to be identical to the variable name, right?

mattiarighi commented:
Please merge the `version2_development` into this branch, this should solve the issues with the pyhton tests.

mattiarighi commented on esmvaltool/recipes/recipe_landcover.yml:
```diff
@@ -0,0 +1,70 @@
+    - loew_al
+    - muel_bn
+    - stac_to
+
```
Please add a maintainer, e.g.:
```
  maintainer:
    righ_ma
```

mattiarighi commented on esmvaltool/recipes/recipe_landcover.yml:
```diff
@@ -0,0 +1,70 @@
+# namelist_landcover.yml
```
Please add `# ESMValTool' as very first line.

mattiarighi commented on esmvaltool/recipes/recipe_landcover.yml:
```diff
@@ -0,0 +1,70 @@
+# namelist_landcover.yml
+---
+documentation:
+  description:
```
Replace with `description: |`

mattiarighi commented:
Thanks @tobstac.
Ok to merge from by side, unless @zklaus or @bouweandela have further remarks.

tobstac commented:
@mattiarighi, thanks  for your help. It made working on this much easier!

bouweandela commented on esmvaltool/diag_scripts/landcover/landcover.py:
```diff
@@ -0,0 +1,507 @@
+    var : str
+        variable short name
+    """
+    from matplotlib.backends.backend_pdf import PdfPages
```
Please put all import statements at the top of the file

bouweandela commented on esmvaltool/preprocessor/_regrid.py:
```diff
@@ -197,6 +197,17 @@ def regrid(cube, target_grid, scheme):
                 [coord] = coords
                 cube.remove_coord(coord)
 
+    # Guess bounds for area weighted regridding if necessary
```
@jvegasbsc Can you comment on this? I though cmorized data always contains bounds?

bouweandela commented on esmvaltool/diag_scripts/landcover/landcover.py:
```diff
@@ -0,0 +1,530 @@
+import matplotlib.pyplot as plt
+from matplotlib.backends.backend_pdf import PdfPages
+import esmvaltool.diag_scripts.shared as diag
+from esmvaltool.diag_scripts.shared._base import ProvenanceLogger
```
```suggestion
from esmvaltool.diag_scripts.shared import ProvenanceLogger
```

tobstac replied:
Ok, done.

bouweandela commented:
@jvegasbsc Can you have a look at the changes to the regrid preprocessor function?

bouweandela commented:
> but do I need to somehow confirm that I have worked on your requested changes to make github aware of this, or is it enough to have committed something?

Writing a comment that you're done changing things is most informative. GitHub does notify me that things have been changed in this pull request, but there is no way to automatically figure out if you're done making changes or still working on something..

tobstac commented:
I just wanted to ask if there is any remaining issue I have to take care of before merging. As far as I am aware, the diagnostic is finished since quite a while now and I would be happy to tick it off in my to-do list.

mattiarighi commented:
I have tested it already, but not with v2 data. So it might be worthy to test it again.
@tobstac just let me know when you are ready.

tobstac commented:
I merged the 'version2_development' into my branch, but can't get it working again. Following the merge I run `python setup.py install`. However, running my diagnostic fails (after pre-processing) with the message
```
[...] ../libstdc++.so.6: version `CXXABI_1.3.9' not found [...]
```
(full logs: diagnostic [log.txt](https://github.com/ESMValGroup/ESMValTool/files/2945140/log.txt) and main [main_log_debug.txt](https://github.com/ESMValGroup/ESMValTool/files/2945170/main_log_debug.txt) ).
I also  run `conda env update` and `pip install -e '.[develop]'` (still not sure whether this is actually the same as doing the `python setup.py install`) but this resulted in the same error.
I guess, something in my conda installation might be broken. Do you have any idea how to check and fix it?

bouweandela commented:
Yes, just update your conda and do a completely fresh installation, following the instructions in CONTRIBUTING.md. This is probably because the dependencies of ESMValTool have been updated.

tobstac commented:
I updated the branch, added provenance tracking to the ascii output files and reverted my changes to _regrid.py. @mattiarighi could you test it, please?

mattiarighi commented:
It works fine, thanks!

Once @bouweandela approves we can merge this.

bouweandela commented on doc/sphinx/source/recipes/recipe_landcover.rst:
```diff
@@ -0,0 +1,127 @@
+   * reference_dataset: land cover extent dataset for comparison. The script was developed using
+     ESACCI-LANDCOVER observations.
+
+   *Optional settings for variables*
```
```suggestion
   *Optional settings for script*
```

tobstac replied:
Done.

bouweandela commented on doc/sphinx/source/recipes/recipe_landcover.rst:
```diff
@@ -0,0 +1,127 @@
+User settings
+-------------
+
+landcover.yml
```
```suggestion
Script landcover.py
```

tobstac replied:
Done, thanks a lot for reviewing!
